src/main/java/com/redhat/devtools/lsp4ij/features/completion/LSPCompletionContributor.java

```diff
@@ -139,6 +139,12 @@ private void addCompletionItems(@NotNull CompletionParameters parameters,
                completionFeature.addLookupItem(completionPrefix, result, lookupItem, size- i, item);
             }
         }
+
+        // If completions were added from LSP and this is auto-completion or the explicit completion trigger was typed
+        // only once, don't show completions from other contributors, specifically the word completion contributor
+        if ((size > 0) && (parameters.getInvocationCount() < 2)) {
```

Comment:
Takes a sample. You are using pycharm and you have a language server which manages python.

In this case completion will not show completion managed with pycharm?

Reply:
As long as the language server returns at least one completion for the current location, only the language server-provided completion(s) will be offered. If there are no completions returned by the language server, any other registered completion contributors are consulted including the word completion contributor.

Let me know if that doesn't make sense, but it's how pretty much all other grammar-aware completion contributors work to provide a proper signal-to-noise ratio.

Reply:
> I misunderstood your example so I'll try again. 

Sorry with my bad explanations. Takes a real sample that we have in IJ Quarkus. Our Quarkus language server contribute to Java language to provide completion snippet (template). When comletion is opened it should show:

 * standard Java completion from IJ
 * the snippet comming from Quarkus language server

If I understand your PR, if snippet from  Quarkus language server is shown, the standard Java completion from IJ will not appear?

> Yes, you're correct that the `LSPCompletionContributor` would show completions from the Python language server and then stop collecting completions before the IDE's native Python plugin. Similarly, here's IntelliJ IDEA Ultimate Edition -- which of course has native TypeScript support -- with the TypeScript language server configured for LSP4IJ:
> 
> ![image](https://private-user-images.githubusercontent.com/7671043/390615785-00f1cbba-97ca-44b2-aa4e-9344ee80cec4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI3NzE2MjMsIm5iZiI6MTczMjc3MTMyMywicGF0aCI6Ii83NjcxMDQzLzM5MDYxNTc4NS0wMGYxY2JiYS05N2NhLTQ0YjItYWE0ZS05MzQ0ZWU4MGNlYzQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTEyOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMjhUMDUyMjAzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NTIzOGIzMTczZTlhN2M4MDg5Y2YxMmVkOGU1NDkwOTY0YmM0YjQ4Mjc3Yzk0N2U2YmVmYzM2ZWQ0NzZlZDM3MSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.R-BXEZ88-uKu12SoPYgS_pijTKHBTrIo7rdhLxJyKxw)
> 
> All of the displayed completions are from LSP4J.
> 
> Here's the same completion request with LSP4IJ's TypeScript language server disabled showing completions from the native TypeScript plugin:
> 
> ![image](https://private-user-images.githubusercontent.com/7671043/390616076-ce4a94b4-bcfa-44b7-9ded-14e4d039b9e5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI3NzE2MjMsIm5iZiI6MTczMjc3MTMyMywicGF0aCI6Ii83NjcxMDQzLzM5MDYxNjA3Ni1jZTRhOTRiNC1iY2ZhLTQ0YjctOWRlZC0xNGU0ZDAzOWI5ZTUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTEyOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMjhUMDUyMjAzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MWI3MTc1Zjc2OTEyYmU3ZTQyOTNlZmM4YzA4M2IwYzQ1MTI0ZGU5MTZiNzhiOTc1ODY2NmVhYTE5MzNlOGIzZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.NB7WdbokQuaB7CRf8Ix8QEsMDWNwzaIv1MgAtABib38)
> 
> I would consider having a language server active for a natively-supported language to be a misconfiguration, though. Is there a compelling use case for that?

Reply:
Ah, yes. That's a compelling use case. It sounds like perhaps disabling other completions should be a per-language client-side setting alongside grammar case-sensitivity. Sounds like the best use of my time immediately is adding those client-side configuration options to unlock some of these other capabilities in a proper manner.

Reply:
I believe this concern should be resolved by the newest commits. Let me know if not.

Reply:
Is your new commit disable native language completion. For instance ifbyiu have ultimate which provides typescript completion and lsp4ij which provides completion with ts languagevserver. What happens ?

If native typescript completion it can be annoying.

I think to avoid doing this choice display native only, display lsp4ij only, display both we could provide a new completion feature method which returns a enum like override_native, ignore_lsp, both.

By default it should show both.

With user definzd ls we could have a setting like casSensitive completion.

What do you think about this idea?

Reply:
No, the most recent change is all about disabling completions from `WordCompletionContributor` if possible by setting `BaseCompletionService.FORBID_WORD_COMPLETION=true` temporarily. The issue here is that `TextMateCompletionContributor` calls the following as its last step:

```java
WordCompletionContributor.addWordCompletionVariants(resultSetWithPrefix, parameters, Collections.emptySet(), true);
```

which muddies the offered completions when LSP4IJ is augmenting a TextMate-managed file type.

Here's an example of TypeScript completions in Ultimate Edition with the TypeScript language server enabled for LSP4IJ:

![image](https://github.com/user-attachments/assets/0e5665a8-a459-4172-82b6-4474afe0c3ad)

As you can see, completions for `Console.log()` have been added from both sources, but you won't see `log` as a word completion here because it's not a TextMate file or even when it is a TextMate file unless you press the completion trigger (**Ctrl+Space**) more than one time.

In the situation where the included change is unable to set `BaseCompletionService.FORBID_WORD_COMPLETION=true` temporarily, it will instead run the remaining completion contributors and filter out redundant completions based on the lookup string, but in actuality, that should never happen because the `CompletionProcess` is always a `UserDataHolder`. Again, that's just me being defensive.

Let me know if that doesn't address your concerns.

Reply:
If your PR disable only word completion when LSP returns somethinf that is excellent!